### DISCUSSION TITLE
Will Cortx R2 support NFS4.1 - Parallel NFS（pNFS）?

### DISCUSSION THREAD

**Discussion by brian479974** [Seagate/cortx#414] [Ideas] [answered]
Will Cortx R2 support NFS4.1 - Parallel NFS（pNFS）?

**Comment by johnbent** [comment, marked as answer]
Hey Brian, thanks for your question!  We are still defining the roadmap.  Probably it won't show up in 2.0 but it may show up in 2.X.  Is a file interface important to you? Is it equally valuable if it is NFS4.0 or do you have a specific workcase requiring 4.1?

**Reply by brian479974** [reply]
**NFS Challenges**
In order to understand how pNFS works it is first necessary to understand what takes place in a typical NFS architecture when a client attempts to access a file. A traditional NFS architecture consists of a filer head placed in front of disk drives and exporting a file system via NFS. When large numbers of clients want to access the data, or if the data set grows too large, the NFS server quickly becomes the bottleneck and significantly impacts system performance because the NFS server sits in the data path between the client computer and the physical storage devices.

**NFS Performance**
pNFS removes the performance bottleneck in traditional NAS systems by allowing the compute clients to read and write data directly and in parallel, to and from the physical storage devices. The NFS server is used only to control metadata and coordinate access, allowing incredibly fast access to very large data sets from many clients.

When a client wants to access a file it first queries the metadata server which provides it with a map of where to find the data and with credentials regarding its rights to read, modify, and write the data. Once the client has those two components, it communicates directly to the storage devices when accessing the data. With traditional NFS every bit of data flows through the NFS server Ã¢Â€Â“ with pNFS the NFS server is removed from the primary data path allowing free and fast access to data. All the advantages of NFS are maintained but bottlenecks are removed and data can be accessed in parallel allowing for very fast throughput rates; system capacity can be easily scaled without impacting overall performance.
**Why is pNFS important?**
pNFS is important because it brings together the benefits of parallel I/O with the benefits of the ubiquitous standard for network file systems (NFS). This will allow users to experience increased performance and scalability in their storage infrastructure with the added assurance that their investment is safe and their ability to choose best-of-breed solutions remains intact.

http://www.pnfs.com/

**Comment by gregnsk** [comment]
@brian479974, NFS support is in our future roadmap. With your comments about pnfs - we don't see much adoption for this NFSv4.1 optional feature in the industry. Do you use pnfs today in production? If you do, would you be available for a short chat about your use case and environment?

**Reply by brian479974** [reply]
@gregnsk Nice to meet you here! 
Which version of NFS will be supported in the future? 
My customer's use case is HPC and they are looking for a scale out solution for their applications. pNFS is not the must but we are interested in learning more form Cortx if there is any good solution for the rapid data growth and performance hungry environment. And we are looking forward to see any cool progress about Cortx!
